CRM/Contact/BAO/Query.php

```diff
@@ -3081,6 +3069,9 @@ public function group($values) {
           $groupIds = implode(',', (array) $smartGroupIDs);
           $gcTable = "civicrm_group_contact_{$this->_groupUniqueKey}";
           $joinClause = ["contact_a.id = {$gcTable}.contact_id"];
+          if ($statusJoinClause) {
+            $joinClause[] = "{$gcTable}.$statusJoinClause";
+          }
           $this->_tables[$gcTable] = $this->_whereTables[$gcTable] = " LEFT JOIN civicrm_group_contact {$gcTable} ON (" . implode(' AND ', $joinClause) . ")";
```

Comment:
This loop and this line cause some queries to have loads and loads of left joins onto the same `civicrm_group_contact` table with unique aliases - I couldn't see a good reason for this and tried to clean it up when we fixed the issue with the AtoZ query but given that it didn't seem to have a big impact on the performance I left that bit out.  But I feel like we only need to actually LEFT JOIN on the table once when the ON joinClause is the same?  Feel free to ignore this comment, but putting it here for reference.

Reply:
@mattwire the only issue is that if you are building a search builder query with 2 separate where clause statements for groups 

e.g. you want to find the overlap between contacts in Group A and contacts in Group B you need to make sure you have separate joins to the group_contact / group_contact_cache tables

Reply:
@mattwire  see dev/core#746

Reply:
@mattwire @seamuslee001 I agree there are more issues here - we just got one logged today

https://lab.civicrm.org/dev/core/issues/930

Once this is merged I might take  a look at further cleanups. I'm encouraged that when I made a mistake here  I got 4 test fails - indicating moderate to good coverage

CRM_Contact_BAO_GroupContactCacheTest.testSmartGroupSearchBuilder
CRM_Contact_BAO_GroupContactCacheTest.testMultipleGroupWhereClause
CRM_Contact_BAO_QueryTest.testGetByGroupWithStatusSmartGroup
CRM_Contact_Form_Search_SearchContactTest.testContactSearchOnGroupType